### PR TITLE
feat: introduce defines for single and double word cpu reg access

### PR DESCRIPTION
**Pull request by mormod**
This is just a quick and dirty substitution of, i.e., `cpu.HL.word` as `CPU_DREG_HL` and `cpu.HL.bytes.high` as `CPU_REG_H`.
This should make the code much more readable and more fool-proof to write. 

Please give me a quick opinion on this.

### CODE REVIEW

File: src/cpu.c [hunk 1]
```diff
@@ -329,242 +329,242 @@ void OPC_LD_SP(void) {
...
 }
 
 void OPC_LD_H_E(void) {
-    LD_REG_REG(&cpu.HL.bytes.high, cpu.DE.bytes.low);
+    LD_REG_REG(&CPU_REG_H, CPU_REG_E);
 }
 
 void OPC_LD_H_H(void) {
-    LD_REG_REG(&cpu.HL.bytes.high, cpu.HL.bytes.high);
+    LD_REG_REG(&CPU_REG_H, cpu.HL.bytes.high);
```
Comment (marcluque) [inline src/cpu.c:488]: I think there's a substitution missing here

File: src/cpu.c [hunk 2]
```diff
@@ -329,242 +329,242 @@ void OPC_LD_SP(void) {
...
 }
 
 void OPC_LD_E_D(void) {
-    LD_REG_REG(&cpu.DE.bytes.low, cpu.DE.bytes.high);
+    LD_REG_REG(&CPU_REG_E, CPU_REG_D);
 }
 
 void OPC_LD_E_E(void) {
-    LD_REG_REG(&cpu.DE.bytes.low, cpu.DE.bytes.low);
+    LD_REG_REG(&CPU_REG_E, cpu.DE.bytes.low);
```
Comment (marcluque) [inline src/cpu.c:450]: Substitution missing

File: src/cpu.c [hunk 3]
```diff
@@ -329,242 +329,242 @@ void OPC_LD_SP(void) {
...
 }
 
 void OPC_LD_D_C(void) {
-    LD_REG_REG(&cpu.DE.bytes.high, cpu.BC.bytes.low);
+    LD_REG_REG(&CPU_REG_D, CPU_REG_C);
 }
 
 void OPC_LD_D_D(void) {
-    LD_REG_REG(&cpu.DE.bytes.high, cpu.DE.bytes.high);
+    LD_REG_REG(&CPU_REG_D, cpu.DE.bytes.high);
```
Comment (marcluque) [inline src/cpu.c:413]: Substitution missing

File: src/cpu.c [hunk 4]
```diff
@@ -329,242 +329,242 @@ void OPC_LD_SP(void) {
...
 }
 
 void OPC_LD_C_B(void) {
-    LD_REG_REG(&cpu.BC.bytes.low, cpu.BC.bytes.high);
+    LD_REG_REG(&CPU_REG_C, CPU_REG_B);
 }
 
 void OPC_LD_C_C(void) {
-    LD_REG_REG(&cpu.BC.bytes.low, cpu.BC.bytes.low);
+    LD_REG_REG(&CPU_REG_C, cpu.BC.bytes.low);
```
Comment (marcluque) [inline src/cpu.c:376]: Substitution missing

File: src/cpu.c [hunk 5]
```diff
@@ -329,242 +329,242 @@ void OPC_LD_SP(void) {
...
-    cpu.HL.word = i-1;
+    uint16_t i = CPU_DREG_HL;
+    mmu_write_byte(i, CPU_REG_A);
+    CPU_DREG_HL = i-1;
     cpu.cycle_count++;
 }
 
 void OPC_LD_B_B(void) {
-    LD_REG_REG(&cpu.BC.bytes.high, cpu.BC.bytes.high);
+    LD_REG_REG(&CPU_REG_B, cpu.BC.bytes.high);
```
Comment (marcluque) [inline src/cpu.c:339]: Substitution missing

File: src/cpu.c [hunk 6]
```diff
@@ -329,242 +329,242 @@ void OPC_LD_SP(void) {
...
 }
 
 void OPC_LD_L_H(void) {
-    LD_REG_REG(&cpu.HL.bytes.low, cpu.HL.bytes.high);
+    LD_REG_REG(&CPU_REG_L, CPU_REG_H);
 }
 
 void OPC_LD_L_L(void) {
-    LD_REG_REG(&cpu.HL.bytes.low, cpu.HL.bytes.low);
+    LD_REG_REG(&CPU_REG_L, cpu.HL.bytes.low);
```
Comment (marcluque) [inline src/cpu.c:525]: Substitution missing

**Comment by marcluque** [comment]
Do you mind (while you're at it) changing the `Register` union?
I think that the naming is not optimal right now since AF is not a register, but rather a register pair, I guess? Do you have any suggestions @mormod @Benzammour?

**Comment by marcluque** [comment]
I think `DoubleWordReg` is a good option since its expressive. Something like `DRegister` could potentially be understood as the D Register sitting in the double word register DE.

**Comment by mormod** [comment]
`DoubleWordRegister` could also reference a register consisting of 2 words, which are not necessarily different registers. 
I would opt for `RegisterPair`.

**Comment by marcluque** [comment]
> `DoubleWordRegister` could also reference a register consisting of 2 words, which are not necessarily different registers. I would opt for `RegisterPair`.

I would have agreed one second ago. I just skimmed through the CPU manual and found this:
image

Hence my answer from above is not correct AF actually is a register, simply not an 8bit register 🤔

File: src/cpu.h [hunk 7]
```diff
@@ -25,19 +25,20 @@ typedef enum flag {
 	Z_FLAG = 7,
 } Flag;
 
-typedef union Register {
+typedef union DoubleWordReg {
 	struct {
 		uint8_t high; // A, B, D, H
 		uint8_t low; // F, C, E, L
 	} bytes;
 	uint16_t word;
```
Comment (marcluque) [inline src/cpu.h:33]: I think this should be called `dword` then, shouldn't it? 🤔

File: src/cpu.h [hunk 8]
```diff
@@ -25,19 +25,20 @@ typedef enum flag {
 	Z_FLAG = 7,
 } Flag;
 
-typedef union Register {
+typedef union DoubleWordReg {
 	struct {
 		uint8_t high; // A, B, D, H
 		uint8_t low; // F, C, E, L
 	} bytes;
```
Comment (marcluque) [inline src/cpu.h:32]: This could be called `words` instead?

File: src/cpu.h (hunk 7 above)
Comment (mormod) [reply src/cpu.h:33]: Yes. `high` and `low` could be renamed to `hi` and `lo` if thats good with you

**Pull request merged**